pkg/controller/cloud/pvlcontroller.go

```diff
@@ -0,0 +1,369 @@
+		// reset things like failure counts for per-item rate limiting
+		pvlc.queue.Forget(key)
+		return true
+	}
+
+	// there was a failure so be sure to report it.  This method allows for pluggable error handling
+	// which can be used for things like cluster-monitoring
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	// since we failed, we should requeue the item to work on later.  This method will add a backoff
+	// to avoid hotlooping on particular items (they're probably still not going to work right away)
+	// and overall controller protection (everything I've done is broken, this controller needs to
+	// calm down or it can starve other useful work) cases.
+	pvlc.queue.AddRateLimited(key)
+
+	return true
+}
+
+// AddLabels adds appropriate labels to persistent volumes and sets the
+// volume as available if successful.
+func (pvlc *PersistentVolumeLabelController) addLabels(key string) error {
+	var volumeLabels map[string]string
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("error getting name of volume %q to get volume from informer: %v", key, err)
+	}
+	volume, err := pvlc.volumeLister.Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("error getting volume %s from informer: %v", name, err)
+	}
+
+	// Only add labels if in the list of initializers
+	if needsInitialization(volume.Initializers, initializerName) {
+		if volume.Spec.AWSElasticBlockStore != nil {
+			labels, err := pvlc.findAWSEBSLabels(volume)
+			if err != nil {
+				return fmt.Errorf("error querying AWS EBS volume %s: %v", volume.Spec.AWSElasticBlockStore.VolumeID, err)
+			}
+			volumeLabels = labels
+		}
+		if volume.Spec.GCEPersistentDisk != nil {
+			labels, err := pvlc.findGCEPDLabels(volume)
+			if err != nil {
+				return fmt.Errorf("error querying GCE PD volume %s: %v", volume.Spec.GCEPersistentDisk.PDName, err)
+			}
+			volumeLabels = labels
+		}
+		return pvlc.updateVolume(volume, volumeLabels)
+	}
+
+	return nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findAWSEBSLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.AWSElasticBlockStore.VolumeID == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+	ebsVolumes, err := pvlc.getEBSVolumes()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: GetVolumeLabels is actually a method on the Volumes interface
+	// If that gets standardized we can refactor to reduce code duplication
+	spec := aws.KubernetesVolumeID(volume.Spec.AWSElasticBlockStore.VolumeID)
+	labels, err := ebsVolumes.GetVolumeLabels(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getEBSVolumes returns the AWS Volumes interface for ebs
+func (pvlc *PersistentVolumeLabelController) getEBSVolumes() (aws.Volumes, error) {
+	pvlc.mutex.Lock()
```

Comment:
followup issue to at least use a RW mutex.  You're only trying this until you succeed once.

Reply:
Created issue #51640